File: article-downloader-image/content/article-downloader.go
package main

import (
	"encoding/json"
	"fmt"
	"hash/fnv"
	"log"
	"os"
	"time"

	"github.com/go-redis/redis"
)

func main() {
	fmt.Println("Article downloader version 0.02")

	agt, pq, lq := getRedisConnections()

	for {
		message := getNextInQueue(lq)

		var link Link
		err := json.Unmarshal([]byte(message), &link)
		if err != nil {
			log.Fatal(err)
		}

		downloadArticle(link, agt, pq)
	}
}

func downloadArticle(link Link, agt *redis.Client, pq *redis.Client) {
	doc, err := GetHTML(link.Url)
	if err != nil {
		fmt.Print("Warning: " + link.Url + " -> ")
		fmt.Print(err)
		fmt.Print("\r\n")
		return
	}

	article, err := GetArticle(doc, link.Url, link.Source)
	if err != nil {
		fmt.Print("Warning: " + link.Url + " -> ")
		fmt.Print(err)
		fmt.Print("\r\n")
		return
	}

	h := hashArticle(article)
	pushed := false

	data, err := json.Marshal(article)
	if err != nil {
		log.Fatal(err)
	}

	if !alreadyGotThat(h, agt) {
		setAlreadyGotThat(h, agt)
		pushNewEntry(string(data), pq)
		pushed = true
	}

	pushedStr := "agt"
	if pushed {
		pushedStr = "new"
	}

	fmt.Println(pushedStr + "\t" + article.Headline)
}

func getNextInQueue(client *redis.Client) string {
	for {
		//fmt.Println("Trying to retrieve message")
		val, err := client.BLPop(60*time.Second, "pending").Result()
		if err == redis.Nil {
			//fmt.Println("No message in queue")
			continue
		} else if err != nil {
			log.Fatal(err)
			time.Sleep(10 * time.Second)
			continue
		} else {
			return val[1]
		}
	}
}

func hashString(s string) uint32 {
	h := fnv.New32a()
	h.Write([]byte(s))
	return h.Sum32()
}

func hashArticle(a Article) uint32 {
	return hashString(a.Headline + a.Content + a.Source)
}

func pushNewEntry(data string, client *redis.Client) {
	err := client.LPush("pending", data).Err()
	if err != nil {
		log.Fatal(err)
	}
}

func alreadyGotThat(hash uint32, client *redis.Client) bool {
	_, err := client.Get(fmt.Sprint(hash)).Result()
	if err == redis.Nil {
		return false
	} else if err != nil {
		log.Fatal(err)
		return true
	} else {
		return true
	}
}

func setAlreadyGotThat(hash uint32, client *redis.Client) {
	expiration := time.Duration(72) * time.Hour
	err := client.Set(fmt.Sprint(hash), "seen", expiration).Err()
	if err != nil {
		log.Fatal(err)
	}
}

func getRedisConnections() (*redis.Client, *redis.Client, *redis.Client) {
	agtUrl := os.Getenv("agt-article-redis-url")
	pqUrl := os.Getenv("pq-redis-url")
	lqUrl := os.Getenv("lq-redis-url")

	if agtUrl == "" || pqUrl == "" || lqUrl == "" {
		log.Fatal("Enviroment variables not set")
	}

	fmt.Println("agt url: " + agtUrl)
	fmt.Println("pq url: " + pqUrl)
	fmt.Println("lq url: " + lqUrl)

	agtClient := redis.NewClient(&redis.Options{
		Addr:     agtUrl + ":6379",
		Password: "",
		DB:       0,
	})

	pqClient := redis.NewClient(&redis.Options{
		Addr:     pqUrl + ":6379",
		Password: "",
		DB:       0,
	})

	lqClient := redis.NewClient(&redis.Options{
		Addr:     lqUrl + ":6379",
		Password: "",
		DB:       0,
	})

	return agtClient, pqClient, lqClient
}

File: article-downloader-image/content/article-downloader.py
import redis
import os
import time as t
from datetime import datetime, date, time, timedelta
import json
from newspaper import Article, Config
from bs4 import BeautifulSoup
import hashlib

def main():
    print('Article downloader version 0.03')

    agtClient, pqClient, lqClient = getRedisConnections()

    while True:
        message = getNextInQueue(lqClient)
        link = json.loads(message[1])
        downloadArticle(url=link['Url'], source=link['Source'], agtClient=agtClient, pqClient=pqClient)

def getRedisConnections():
    agtUrl = os.environ.get('agt-article-redis-url')
    pqUrl = os.environ.get('pq-redis-url')
    lqUrl = os.environ.get('lq-redis-url')

    if agtUrl == '' or pqUrl == '' or lqUrl == '':
        print('Environment variables are not set')

    rp = 6379 #default redis port
    agtClient = redis.Redis(host=agtUrl, port=rp, password='', db=0)
    pqClient = redis.Redis(host=pqUrl, port=rp, password='', db=0)
    lqUrl = redis.Redis(host=lqUrl, port=rp, password='', db=0)
    return agtClient, pqClient, lqUrl

def getNextInQueue(client):
    while True:
        rval = client.blpop(keys='pending', timeout=60)
        if not rval:
            t.sleep(10)
            continue
        return rval

def downloadArticle(url, source, agtClient, pqClient):
    config = Config()
    config.MIN_WORD_COUNT = 100
    article = Article(url=url, config=config)
    try:
        article.download()
        article.parse()
    except:
        return

    if article.meta_lang in ['en']:
        articleHtml = article.html
        soup = BeautifulSoup(articleHtml, 'html.parser')
        description = ''
        try:
            if not description:
                description = soup.find('meta', attrs={'name':'description'}).get('content')
        except:
            if not description:
                description = ''
        try:
            if not description:
                description = soup.find('meta', attrs={'property':'og:description'}).get('content')
        except:
            if not description:
                description = ''
        try:
            if not description:
                description = soup.find('meta', attrs={'property':'twitter:description'}).get('content')
        except:
            if not description:
                description = ''
        a = {
            'Headline': article.title,
            'Description': description,
            'Image': article.top_image,
            'Content': article.text,
            'Source': source,
            'Url': url,
            'Time': str(datetime.now())
        }
        h = hashArticle(a)
        pushed = False
        if not alreadyGotThat(h, agtClient):
            setAlreadyGotThat(h, agtClient)
            data = json.dumps(a)
            pushNewEntry(data, pqClient)
            pushed = True

        pushedMsg = 'agt'
        if pushed:
            pushedMsg = 'new'

        print(pushedMsg, "\t", a)

def hashArticle(a):
    return hashlib.md5(bytes(a['Headline'] + a['Content'] + a['Source'], encoding='utf8')).hexdigest()

def alreadyGotThat(hash, client):
    if client.exists(hash):
        return True
    return False

def setAlreadyGotThat(hash, client):
    expiration = timedelta(hours=72)
    client.set(name=hash, value='seen', ex=expiration)

def pushNewEntry(data, client):
    client.lpush('pending', data)


if __name__ == "__main__":
    main()